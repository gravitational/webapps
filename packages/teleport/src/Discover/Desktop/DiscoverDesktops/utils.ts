import { Path } from 'teleport/Discover/Desktop/DiscoverDesktops/lines/Path';
import { MonotoneX } from 'teleport/Discover/Desktop/DiscoverDesktops/lines/MonotoneX';

const STROKE_WIDTH = 4;

export interface Line {
  width: number;
  height: number;
  path: string;
}

export function createLine(
  desktopServiceElement: HTMLDivElement,
  desktopElement: HTMLDivElement,
  containerElement: HTMLDivElement
): Line {
  if (!desktopElement || !desktopServiceElement || !containerElement) {
    return null;
  }

  const desktopServiceRect = desktopServiceElement.getBoundingClientRect();
  const desktopRect = desktopElement.getBoundingClientRect();
  const containerRect = containerElement.getBoundingClientRect();

  const distance = desktopRect.left - desktopServiceRect.right;

  const path = new Path();
  const line = new MonotoneX(path);

  line.lineStart();

  const desktopLinePosition =
    desktopRect.top - containerRect.top + desktopRect.height / 2 - 1;
  const desktopServiceLinePosition =
    desktopServiceRect.top - containerRect.top + desktopServiceRect.height / 2;

  line.point(0, desktopServiceLinePosition - STROKE_WIDTH * 2);
  line.point(40, desktopServiceLinePosition - STROKE_WIDTH * 2);
  line.point(distance - 10, desktopLinePosition + STROKE_WIDTH / 2);
  line.point(distance, desktopLinePosition + STROKE_WIDTH / 2);

  line.lineEnd();

  return {
    width: distance,
    height: containerRect.height,
    path: path.path,
  };
}
