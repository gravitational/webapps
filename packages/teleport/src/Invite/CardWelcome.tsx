/*
Copyright 2021 Gravitational, Inc.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

import React from 'react';
import { Card, Box, Text, ButtonPrimary } from 'design';
import cfg from 'teleport/config';
import history from 'teleport/services/history';

export default function CardWelcome({ tokenId, inviteMode = true }: Props) {
  const title = inviteMode ? 'Welcome to Teleport' : 'Reset Password';
  const description = inviteMode
    ? 'Please click the button below to create an account'
    : 'Please click the button below to begin recovery of your account';
  const buttonText = inviteMode ? 'Get started' : 'Continue';

  const nextStepPath = inviteMode
    ? cfg.getUserInviteTokenContinueRoute(tokenId)
    : cfg.getUserResetTokenContinueRoute(tokenId);

  return (
    <Card bg="primary.light" my={6} mx="auto" width="464px">
      <Box p={6}>
        <Text typography="h2" mb={3} textAlign="center" color="light">
          {title}
        </Text>
        <Text typography="h5" mb={3} textAlign="center">
          {description}
        </Text>
        <ButtonPrimary
          width="100%"
          mt={3}
          size="large"
          onClick={() => history.push(nextStepPath)}
        >
          {buttonText}
        </ButtonPrimary>
      </Box>
    </Card>
  );
}

type Props = {
  tokenId: string;
  inviteMode?: boolean;
};
