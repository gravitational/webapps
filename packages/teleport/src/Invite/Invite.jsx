/*
Copyright 2020 Gravitational, Inc.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

import React from 'react';
import { useParams } from 'react-router';
import { useAttempt, withState } from 'shared/hooks';
import cfg from 'teleport/config';
import auth from 'teleport/services/auth';
import history from 'teleport/services/history';
import Logger from 'shared/libs/logger';
import InviteForm, { Expired } from 'shared/components/FormInvite';
import LogoHero from 'teleport/components/LogoHero';

const logger = Logger.create('components/Invite');

export function Invite(props) {
  const {
    passwordResetMode = false,
    auth2faType,
    fetchAttempt,
    onSubmit,
    onSubmitWithU2f,
    passwordToken,
    submitAttempt,
  } = props;

  if (fetchAttempt.isFailed) {
    return (
      <>
        <LogoHero />
        <Expired />
      </>
    );
  }

  if (!fetchAttempt.isSuccess) {
    return null;
  }

  const { user, qrCode } = passwordToken;
  const title = passwordResetMode ? 'Reset Password' : 'Welcome to Teleport';
  const submitBtnText = passwordResetMode
    ? 'Change Password'
    : 'Create Account';

  return (
    <>
      <LogoHero />
      <InviteForm
        submitBtnText={submitBtnText}
        title={title}
        user={user}
        qr={qrCode}
        auth2faType={auth2faType}
        attempt={submitAttempt}
        onSubmitWithU2f={onSubmitWithU2f}
        onSubmit={onSubmit}
      />
    </>
  );
}

function mapState() {
  const [fetchAttempt, fetchAttemptActions] = useAttempt();
  const [passwordToken, setPswToken] = React.useState();
  const [submitAttempt, submitAttemptActions] = useAttempt();
  const auth2faType = cfg.getAuth2faType();
  const { tokenId } = useParams();

  React.useEffect(() => {
    fetchAttemptActions.do(() => {
      return auth
        .fetchPasswordToken(tokenId)
        .then(resetToken => setPswToken(resetToken));
    });
  }, []);

  function redirect() {
    history.push(cfg.routes.root, true);
  }

  function handleSubmitError(err) {
    logger.error(err);
    submitAttemptActions.error(err);
  }

  function onSubmit(password, otpToken) {
    submitAttemptActions.start();
    auth
      .resetPassword(tokenId, password, otpToken)
      .then(redirect)
      .catch(handleSubmitError);
  }

  function onSubmitWithU2f(password) {
    submitAttemptActions.start();
    auth
      .resetPasswordWithU2f(tokenId, password)
      .then(redirect)
      .catch(handleSubmitError);
  }

  return {
    auth2faType,
    fetchAttempt,
    onSubmit,
    onSubmitWithU2f,
    passwordToken,
    submitAttempt,
    tokenId,
  };
}

const InviteWithState = withState(mapState)(Invite);

export default InviteWithState;
export const ResetPassword = () => <InviteWithState passwordResetMode={true} />;
