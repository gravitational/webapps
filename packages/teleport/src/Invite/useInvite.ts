/*
Copyright 2021 Gravitational, Inc.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

import { useState, useEffect } from 'react';
import useAttempt from 'shared/hooks/useAttemptNext';
import cfg from 'teleport/config';
import history from 'teleport/services/history';
import auth from 'teleport/services/auth';

export default function useInvite(tokenId: string) {
  const [passwordToken, setPswToken] = useState<ResetToken>();
  const fetchAttempt = useAttempt('');
  const submitAttempt = useAttempt('');
  const auth2faType = cfg.getAuth2faType();

  useEffect(() => {
    fetchAttempt.run(() =>
      auth
        .fetchPasswordToken(tokenId)
        .then(resetToken => setPswToken(resetToken))
    );
  }, []);

  function onSubmit(password: string, otpToken: string) {
    submitAttempt.setAttempt({ status: 'processing' });
    auth
      .resetPassword(tokenId, password, otpToken)
      .then(redirect)
      .catch(submitAttempt.handleError);
  }

  function onSubmitWithU2f(password: string) {
    submitAttempt.setAttempt({ status: 'processing' });
    auth
      .resetPasswordWithU2f(tokenId, password)
      .then(redirect)
      .catch(submitAttempt.handleError);
  }

  function onSubmitWithWebauthn(password: string) {
    submitAttempt.setAttempt({ status: 'processing' });
    auth
      .resetPasswordWithWebauthn(tokenId, password)
      .then(redirect)
      .catch(submitAttempt.handleError);
  }

  function redirect() {
    history.push(cfg.routes.root, true);
  }

  function clearSubmitAttempt() {
    submitAttempt.setAttempt({ status: '' });
  }

  return {
    auth2faType,
    preferredMfaType: cfg.getPreferredMfaType(),
    fetchAttempt: fetchAttempt.attempt,
    submitAttempt: submitAttempt.attempt,
    clearSubmitAttempt,
    onSubmit,
    onSubmitWithU2f,
    onSubmitWithWebauthn,
    passwordToken,
  };
}

type ResetToken = {
  tokenId: string;
  qrCode: string;
  user: string;
};

export type State = ReturnType<typeof useInvite>;
