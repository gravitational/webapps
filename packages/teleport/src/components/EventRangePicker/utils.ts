import moment from 'moment';

export function getRangeOptions(): EventRange[] {
  return [
    {
      name: 'Today',
      from: moment(new Date())
        .startOf('day')
        .toDate(),
      to: moment(new Date())
        .endOf('day')
        .toDate(),
    },
    {
      name: '7 days',
      from: moment()
        .subtract(6, 'day')
        .startOf('day')
        .toDate(),
      to: moment(new Date())
        .endOf('day')
        .toDate(),
    },
    {
      name: 'Custom Range...',
      isCustom: true,
      from: new Date(),
      to: new Date(),
    },
  ];
}

export type EventRange = {
  from: Date;
  to: Date;
  isCustom?: boolean;
  name?: string;
};
