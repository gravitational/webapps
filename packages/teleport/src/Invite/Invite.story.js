/*
Copyright 2020 Gravitational, Inc.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

import React from 'react';
import { Invite as Component } from './Invite';

export default {
  title: 'Teleport/Invite',
  component: Component,
};

export function Invite() {
  return <Component {...defaultProps} />;
}

export function Expired() {
  const props = {
    ...defaultProps,
    fetchAttempt: { isFailed: true, message: 'this is error message' },
  };

  return <Component {...props} />;
}

export function ResetPasswordScreen() {
  const props = {
    ...defaultProps,
    passwordResetMode: true,
  };
  return <Component {...props} />;
}

export function AuthTfaOn() {
  const props = {
    ...defaultProps,
    auth2faType: 'on',
  };
  return <Component {...props} />;
}

export function AuthTfaOptional() {
  const props = {
    ...defaultProps,
    auth2faType: 'optional',
  };
  return <Component {...props} />;
}

const defaultProps = {
  auth2faType: 'off',
  submitAttempt: {},
  fetchAttempt: {
    isSuccess: true,
  },
  onSubmitWithU2f: () => null,
  fetchUserToken: () => null,
  onSubmit: () => null,
  passwordToken: {
    user: 'john@example.com',
    url: 'https://localhost/sampleurl',
  },
};
