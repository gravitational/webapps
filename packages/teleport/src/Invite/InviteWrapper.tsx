import React from 'react';
import cfg from 'teleport/config';
import { Route, Switch, useParams } from 'teleport/components/Router';
import LogoHero from 'teleport/components/LogoHero';
import CardWelcome from './CardWelcome';

export default function InviteWrapper({ passwordResetMode = false, children }) {
  const { tokenId } = useParams<{ tokenId: string }>();

  return (
    <>
      <LogoHero />
      <Switch>
        <Route
          path={[cfg.routes.userInviteContinue, cfg.routes.userResetContinue]}
        >
          {children}
        </Route>
        <Route path={[cfg.routes.userInvite, cfg.routes.userReset]}>
          <CardWelcome
            tokenId={tokenId}
            passwordResetMode={passwordResetMode}
          />
        </Route>
      </Switch>
    </>
  );
}
