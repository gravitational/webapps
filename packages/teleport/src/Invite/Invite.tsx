/*
Copyright 2021 Gravitational, Inc.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

import React from 'react';
import { useParams } from 'react-router';
import InviteForm, { Expired } from 'shared/components/FormInvite';
import LogoHero from 'teleport/components/LogoHero';
import useInvite, { State } from './useInvite';

export default function Container({ passwordResetMode = false }) {
  const { tokenId } = useParams<{ tokenId: string }>();
  const state = useInvite(tokenId);
  return <Invite {...state} passwordResetMode={passwordResetMode} />;
}

export function Invite(props: State & Props) {
  const {
    passwordResetMode,
    auth2faType,
    fetchAttempt,
    submitAttempt,
    clearSubmitAttempt,
    onSubmit,
    onSubmitWithU2f,
    passwordToken,
  } = props;

  if (fetchAttempt.status === 'failed') {
    return (
      <>
        <LogoHero />
        <Expired />
      </>
    );
  }

  if (fetchAttempt.status !== 'success') {
    return null;
  }

  const { user, qrCode } = passwordToken;
  const title = passwordResetMode ? 'Reset Password' : 'Welcome to Teleport';
  const submitBtnText = passwordResetMode
    ? 'Change Password'
    : 'Create Account';

  return (
    <>
      <LogoHero />
      <InviteForm
        submitBtnText={submitBtnText}
        title={title}
        user={user}
        qr={qrCode}
        auth2faType={auth2faType}
        attempt={submitAttempt}
        clearSubmitAttempt={clearSubmitAttempt}
        onSubmitWithU2f={onSubmitWithU2f}
        onSubmit={onSubmit}
      />
    </>
  );
}

export type Props = {
  passwordResetMode: boolean;
};

export const ResetPassword = () => <Container passwordResetMode={true} />;
