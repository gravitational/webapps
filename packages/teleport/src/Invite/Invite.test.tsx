/*
Copyright 2021 Gravitational, Inc.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

import React from 'react';
import { MemoryRouter, Route } from 'react-router';
import { screen, fireEvent, act, render } from 'design/utils/testing';
import { Logger } from 'shared/libs/logger';
import cfg from 'teleport/config';
import auth from 'teleport/services/auth';
import Invite from './Invite';
import { AuthMfaOn, AuthMfaOptional } from './Invite.story';

describe('teleport/components/Invite', () => {
  beforeEach(() => {
    jest.spyOn(Logger.prototype, 'log').mockImplementation();
    jest.spyOn(auth, 'fetchPasswordToken').mockImplementation(async () => ({
      user: 'sam',
      tokenId: 'test123',
      qrCode: 'test12345',
    }));
  });

  afterEach(() => {
    jest.clearAllMocks();
  });

  it('should show "get started" prompt before showing form', async () => {
    await act(async () =>
      render(
        <MemoryRouter initialEntries={['/web/invite/5182']}>
          <Invite />
        </MemoryRouter>
      )
    );

    expect(screen.getByText(/get started/i)).toBeInTheDocument();
  });

  it('reset password', async () => {
    jest.spyOn(cfg, 'getAuth2faType').mockImplementation(() => 'off');
    jest
      .spyOn(auth, 'resetPassword')
      .mockImplementation(() => new Promise(() => null));

    await act(async () => renderInvite());

    const pwdField = screen.getByPlaceholderText('Password');
    const pwdConfirmField = screen.getByPlaceholderText('Confirm Password');

    // fill out input boxes and trigger submit
    fireEvent.change(pwdField, { target: { value: 'pwd_value' } });
    fireEvent.change(pwdConfirmField, { target: { value: 'pwd_value' } });
    fireEvent.click(screen.getByRole('button'));

    expect(auth.resetPassword).toHaveBeenCalledWith('5182', 'pwd_value', '');
  });

  it('reset password with otp', async () => {
    jest.spyOn(cfg, 'getAuth2faType').mockImplementation(() => 'otp');
    jest
      .spyOn(auth, 'resetPassword')
      .mockImplementation(() => new Promise(() => null));

    await act(async () => renderInvite());

    const pwdField = screen.getByPlaceholderText('Password');
    const pwdConfirmField = screen.getByPlaceholderText('Confirm Password');
    const otpField = screen.getByPlaceholderText('123 456');

    // fill out input boxes and trigger submit
    fireEvent.change(pwdField, { target: { value: 'pwd_value' } });
    fireEvent.change(pwdConfirmField, { target: { value: 'pwd_value' } });
    fireEvent.change(otpField, { target: { value: '2222' } });
    fireEvent.click(screen.getByRole('button'));

    expect(auth.resetPassword).toHaveBeenCalledWith(
      '5182',
      'pwd_value',
      '2222'
    );
  });

  it('reset password with U2F', async () => {
    jest.spyOn(cfg, 'getAuth2faType').mockImplementation(() => 'u2f');
    jest
      .spyOn(auth, 'resetPassword')
      .mockImplementation(() => new Promise(() => null));
    jest
      .spyOn(auth, 'resetPasswordWithU2f')
      .mockImplementation(() => new Promise(() => null));

    await act(async () => renderInvite());

    // fill out input boxes and trigger submit
    const pwdField = screen.getByPlaceholderText('Password');
    const pwdConfirmField = screen.getByPlaceholderText('Confirm Password');
    fireEvent.change(pwdField, { target: { value: 'pwd_value' } });
    fireEvent.change(pwdConfirmField, { target: { value: 'pwd_value' } });
    fireEvent.click(screen.getByRole('button'));

    expect(auth.resetPassword).not.toHaveBeenCalled();
    expect(auth.resetPasswordWithU2f).toHaveBeenCalledWith('5182', 'pwd_value');
  });

  it('reset password with webauthn', async () => {
    jest.spyOn(cfg, 'getAuth2faType').mockImplementation(() => 'webauthn');
    jest
      .spyOn(auth, 'resetPassword')
      .mockImplementation(() => new Promise(() => null));
    jest
      .spyOn(auth, 'resetPasswordWithWebauthn')
      .mockImplementation(() => new Promise(() => null));

    await act(async () => renderInvite());

    // fill out input boxes and trigger submit
    const pwdField = screen.getByPlaceholderText('Password');
    const pwdConfirmField = screen.getByPlaceholderText('Confirm Password');
    fireEvent.change(pwdField, { target: { value: 'pwd_value' } });
    fireEvent.change(pwdConfirmField, { target: { value: 'pwd_value' } });
    fireEvent.click(screen.getByRole('button'));

    expect(auth.resetPasswordWithWebauthn).toHaveBeenCalledWith(
      '5182',
      'pwd_value'
    );
  });

  it('reset password error', async () => {
    let reject;
    jest.spyOn(cfg, 'getAuth2faType').mockImplementation(() => 'off');
    jest.spyOn(auth, 'resetPassword').mockImplementation(() => {
      return new Promise((resolve, _reject) => {
        reject = _reject;
      });
    });

    await act(async () => renderInvite());

    // fill out input boxes and trigger submit
    const pwdField = screen.getByPlaceholderText('Password');
    const pwdConfirmField = screen.getByPlaceholderText('Confirm Password');
    fireEvent.change(pwdField, { target: { value: 'pwd_value' } });
    fireEvent.change(pwdConfirmField, { target: { value: 'pwd_value' } });

    await act(async () => {
      fireEvent.click(screen.getByRole('button'));
      reject(new Error('server_error'));
    });

    expect(screen.getByText('server_error')).toBeDefined();
  });

  it('auth type "on" should render form with hardware key as first option in dropdown', () => {
    const { container } = render(
      <MemoryRouter initialEntries={['/web/invite/5182/continue']}>
        <AuthMfaOn />
      </MemoryRouter>
    );
    expect(container).toMatchSnapshot();
  });

  it('auth type "optional" should render form with hardware key as first option in dropdown', () => {
    const { container } = render(
      <MemoryRouter initialEntries={['/web/invite/5182/continue']}>
        <AuthMfaOptional />
      </MemoryRouter>
    );
    expect(container).toMatchSnapshot();
  });
});

function renderInvite(url = `/web/invite/5182/continue`) {
  render(
    <MemoryRouter initialEntries={[url]}>
      <Route path={cfg.routes.userInviteContinue}>
        <Invite />
      </Route>
    </MemoryRouter>
  );
}
