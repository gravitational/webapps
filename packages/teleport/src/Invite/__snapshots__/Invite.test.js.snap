// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`teleport/components/Invite should render form with token requirement when auth2fatype is on 1`] = `
.c3 {
  box-sizing: border-box;
  padding: 40px;
  flex: 3;
}

.c6 {
  box-sizing: border-box;
  margin-bottom: 24px;
}

.c10 {
  box-sizing: border-box;
  margin-bottom: 24px;
  width: 50%;
}

.c12 {
  box-sizing: border-box;
  padding: 40px;
  background-color: #1C254D;
  flex: 1;
  border-top-right-radius: 8px;
  border-bottom-right-radius: 8px;
}

.c11 {
  line-height: 1.5;
  margin: 0;
  display: inline-flex;
  justify-content: center;
  align-items: center;
  box-sizing: border-box;
  border: none;
  border-radius: 4px;
  cursor: pointer;
  font-family: inherit;
  font-weight: 600;
  outline: none;
  position: relative;
  text-align: center;
  text-decoration: none;
  text-transform: uppercase;
  transition: all 0.3s;
  -webkit-font-smoothing: antialiased;
  background: #512FC9;
  color: rgba(255,255,255,0.87);
  min-height: 40px;
  font-size: 12px;
  padding: 0px 40px;
  margin-top: 16px;
  width: 100%;
}

.c11:active {
  opacity: 0.56;
}

.c11:hover,
.c11:focus {
  background: #651FFF;
}

.c11:active {
  background: #354AA4;
}

.c11:disabled {
  background: rgba(255,255,255,0.12);
  color: rgba(255,255,255,0.3);
}

.c15 {
  line-height: 1.5;
  margin: 0;
  display: inline-flex;
  justify-content: center;
  align-items: center;
  box-sizing: border-box;
  border: none;
  border-radius: 4px;
  cursor: pointer;
  font-family: inherit;
  font-weight: 600;
  outline: none;
  position: relative;
  text-align: center;
  text-decoration: none;
  text-transform: uppercase;
  transition: all 0.3s;
  -webkit-font-smoothing: antialiased;
  background: #222C59;
  color: rgba(255,255,255,0.87);
  font-size: 10px;
  min-height: 24px;
  padding: 0px 16px;
  width: 100%;
}

.c15:active {
  opacity: 0.56;
}

.c15:hover,
.c15:focus {
  background: #2C3A73;
}

.c15:disabled {
  background: rgba(255,255,255,0.12);
  color: rgba(255,255,255,0.3);
}

.c14 {
  color: #03a9f4;
  font-weight: normal;
  background: none;
  text-decoration: underline;
  text-transform: none;
  padding: 0 8px;
}

.c14:hover,
.c14:focus {
  background: #222C59;
}

.c1 {
  box-sizing: border-box;
  margin-left: auto;
  margin-right: auto;
  margin-top: 40px;
  margin-bottom: 40px;
  width: 720px;
  background-color: #222C59;
  box-shadow: 0 4px 16px rgba(0,0,0,0.24);
  border-radius: 8px;
}

.c8 {
  appearance: none;
  border: none;
  border-radius: 4px;
  box-shadow: inset 0 2px 4px rgba(0,0,0,.24);
  box-sizing: border-box;
  display: block;
  height: 40px;
  font-size: 16px;
  padding: 0 16px;
  outline: none;
  width: 100%;
  color: #324148;
  background-color: #FFFFFF;
}

.c8::-ms-clear {
  display: none;
}

.c8::placeholder {
  opacity: 0.4;
}

.c8:read-only {
  cursor: not-allowed;
}

.c7 {
  color: #FFFFFF;
  display: block;
  font-size: 11px;
  font-weight: 500;
  text-transform: uppercase;
  width: 100%;
  margin-bottom: 4px;
}

.c0 {
  display: block;
  outline: none;
  margin-left: auto;
  margin-right: auto;
  margin-top: 40px;
  margin-bottom: 40px;
  max-width: 200px;
  max-height: 120px;
}

.c4 {
  overflow: hidden;
  text-overflow: ellipsis;
  font-weight: 300;
  font-size: 28px;
  line-height: 32px;
  margin: 0px;
  margin-bottom: 16px;
  color: #FFFFFF;
  text-align: center;
}

.c5 {
  overflow: hidden;
  text-overflow: ellipsis;
  font-weight: 400;
  font-size: 18px;
  line-height: 32px;
  word-break: break-all;
  margin: 0px;
  margin-bottom: 16px;
}

.c13 {
  overflow: hidden;
  text-overflow: ellipsis;
  font-weight: 300;
  font-size: 12px;
  line-height: 24px;
  margin: 0px;
  margin-bottom: 16px;
}

.c2 {
  box-sizing: border-box;
  display: flex;
}

.c9 {
  box-sizing: border-box;
  display: flex;
  flex-direction: row;
}

<div>
  <img
    class="c0"
    src="file_stub"
  />
  <form
    class="c1"
    width="720px"
  >
    <div
      class="c2"
    >
      <div
        class="c3"
      >
        <div
          class="c4"
          color="light"
        >
          Welcome to Teleport
        </div>
        <div
          class="c5"
        >
          john@example.com
        </div>
        <div
          class="c6"
        >
          <label
            class="c7"
            font-size="0"
          >
            Password
          </label>
          <input
            autocomplete="off"
            class="c8"
            color="text.onLight"
            placeholder="Password"
            type="password"
            value=""
          />
        </div>
        <div
          class="c6"
        >
          <label
            class="c7"
            font-size="0"
          >
            Confirm Password
          </label>
          <input
            autocomplete="off"
            class="c8"
            color="text.onLight"
            placeholder="Confirm Password"
            type="password"
            value=""
          />
        </div>
        <div
          class="c9"
        >
          <div
            class="c10"
            width="50%"
          >
            <label
              class="c7"
              font-size="0"
            >
              Two factor token
            </label>
            <input
              autocomplete="off"
              class="c8"
              color="text.onLight"
              placeholder="123 456"
              type="text"
              value=""
            />
          </div>
        </div>
        <button
          class="c11"
          kind="primary"
          width="100%"
        >
          Create Account
        </button>
      </div>
      <div
        class="c12"
      >
        <div>
          <div
            class="c13"
          >
            Scan the bar code with Google Authenticator to generate a two factor token.
          </div>
          <img
            src="data:image/png;base64,undefined"
            style="border: 8px solid;"
            width="152"
          />
          <a
            class="c14 c15"
            href="https://support.google.com/accounts/answer/1066447?co=GENIE.Platform%3DiOS&hl=en&oco=0"
            kind="secondary"
            rel="noreferrer"
            target="_blank"
            width="100%"
          >
            Download Google Authenticator
          </a>
        </div>
      </div>
    </div>
  </form>
</div>
`;

exports[`teleport/components/Invite should render form with token requirement when auth2fatype is optional 1`] = `
.c3 {
  box-sizing: border-box;
  padding: 40px;
  flex: 3;
}

.c6 {
  box-sizing: border-box;
  margin-bottom: 24px;
}

.c10 {
  box-sizing: border-box;
  margin-bottom: 24px;
  width: 50%;
}

.c12 {
  box-sizing: border-box;
  padding: 40px;
  background-color: #1C254D;
  flex: 1;
  border-top-right-radius: 8px;
  border-bottom-right-radius: 8px;
}

.c11 {
  line-height: 1.5;
  margin: 0;
  display: inline-flex;
  justify-content: center;
  align-items: center;
  box-sizing: border-box;
  border: none;
  border-radius: 4px;
  cursor: pointer;
  font-family: inherit;
  font-weight: 600;
  outline: none;
  position: relative;
  text-align: center;
  text-decoration: none;
  text-transform: uppercase;
  transition: all 0.3s;
  -webkit-font-smoothing: antialiased;
  background: #512FC9;
  color: rgba(255,255,255,0.87);
  min-height: 40px;
  font-size: 12px;
  padding: 0px 40px;
  margin-top: 16px;
  width: 100%;
}

.c11:active {
  opacity: 0.56;
}

.c11:hover,
.c11:focus {
  background: #651FFF;
}

.c11:active {
  background: #354AA4;
}

.c11:disabled {
  background: rgba(255,255,255,0.12);
  color: rgba(255,255,255,0.3);
}

.c15 {
  line-height: 1.5;
  margin: 0;
  display: inline-flex;
  justify-content: center;
  align-items: center;
  box-sizing: border-box;
  border: none;
  border-radius: 4px;
  cursor: pointer;
  font-family: inherit;
  font-weight: 600;
  outline: none;
  position: relative;
  text-align: center;
  text-decoration: none;
  text-transform: uppercase;
  transition: all 0.3s;
  -webkit-font-smoothing: antialiased;
  background: #222C59;
  color: rgba(255,255,255,0.87);
  font-size: 10px;
  min-height: 24px;
  padding: 0px 16px;
  width: 100%;
}

.c15:active {
  opacity: 0.56;
}

.c15:hover,
.c15:focus {
  background: #2C3A73;
}

.c15:disabled {
  background: rgba(255,255,255,0.12);
  color: rgba(255,255,255,0.3);
}

.c14 {
  color: #03a9f4;
  font-weight: normal;
  background: none;
  text-decoration: underline;
  text-transform: none;
  padding: 0 8px;
}

.c14:hover,
.c14:focus {
  background: #222C59;
}

.c1 {
  box-sizing: border-box;
  margin-left: auto;
  margin-right: auto;
  margin-top: 40px;
  margin-bottom: 40px;
  width: 720px;
  background-color: #222C59;
  box-shadow: 0 4px 16px rgba(0,0,0,0.24);
  border-radius: 8px;
}

.c8 {
  appearance: none;
  border: none;
  border-radius: 4px;
  box-shadow: inset 0 2px 4px rgba(0,0,0,.24);
  box-sizing: border-box;
  display: block;
  height: 40px;
  font-size: 16px;
  padding: 0 16px;
  outline: none;
  width: 100%;
  color: #324148;
  background-color: #FFFFFF;
}

.c8::-ms-clear {
  display: none;
}

.c8::placeholder {
  opacity: 0.4;
}

.c8:read-only {
  cursor: not-allowed;
}

.c7 {
  color: #FFFFFF;
  display: block;
  font-size: 11px;
  font-weight: 500;
  text-transform: uppercase;
  width: 100%;
  margin-bottom: 4px;
}

.c0 {
  display: block;
  outline: none;
  margin-left: auto;
  margin-right: auto;
  margin-top: 40px;
  margin-bottom: 40px;
  max-width: 200px;
  max-height: 120px;
}

.c4 {
  overflow: hidden;
  text-overflow: ellipsis;
  font-weight: 300;
  font-size: 28px;
  line-height: 32px;
  margin: 0px;
  margin-bottom: 16px;
  color: #FFFFFF;
  text-align: center;
}

.c5 {
  overflow: hidden;
  text-overflow: ellipsis;
  font-weight: 400;
  font-size: 18px;
  line-height: 32px;
  word-break: break-all;
  margin: 0px;
  margin-bottom: 16px;
}

.c13 {
  overflow: hidden;
  text-overflow: ellipsis;
  font-weight: 300;
  font-size: 12px;
  line-height: 24px;
  margin: 0px;
  margin-bottom: 16px;
}

.c2 {
  box-sizing: border-box;
  display: flex;
}

.c9 {
  box-sizing: border-box;
  display: flex;
  flex-direction: row;
}

<div>
  <img
    class="c0"
    src="file_stub"
  />
  <form
    class="c1"
    width="720px"
  >
    <div
      class="c2"
    >
      <div
        class="c3"
      >
        <div
          class="c4"
          color="light"
        >
          Welcome to Teleport
        </div>
        <div
          class="c5"
        >
          john@example.com
        </div>
        <div
          class="c6"
        >
          <label
            class="c7"
            font-size="0"
          >
            Password
          </label>
          <input
            autocomplete="off"
            class="c8"
            color="text.onLight"
            placeholder="Password"
            type="password"
            value=""
          />
        </div>
        <div
          class="c6"
        >
          <label
            class="c7"
            font-size="0"
          >
            Confirm Password
          </label>
          <input
            autocomplete="off"
            class="c8"
            color="text.onLight"
            placeholder="Confirm Password"
            type="password"
            value=""
          />
        </div>
        <div
          class="c9"
        >
          <div
            class="c10"
            width="50%"
          >
            <label
              class="c7"
              font-size="0"
            >
              Two factor token
            </label>
            <input
              autocomplete="off"
              class="c8"
              color="text.onLight"
              placeholder="123 456"
              type="text"
              value=""
            />
          </div>
        </div>
        <button
          class="c11"
          kind="primary"
          width="100%"
        >
          Create Account
        </button>
      </div>
      <div
        class="c12"
      >
        <div>
          <div
            class="c13"
          >
            Scan the bar code with Google Authenticator to generate a two factor token.
          </div>
          <img
            src="data:image/png;base64,undefined"
            style="border: 8px solid;"
            width="152"
          />
          <a
            class="c14 c15"
            href="https://support.google.com/accounts/answer/1066447?co=GENIE.Platform%3DiOS&hl=en&oco=0"
            kind="secondary"
            rel="noreferrer"
            target="_blank"
            width="100%"
          >
            Download Google Authenticator
          </a>
        </div>
      </div>
    </div>
  </form>
</div>
`;
